Fuzz testing failed

**Version:** latest master branch of [indradb/indradb](https://github.com/indradb/indradb)
**Env:** archlinux + AMD 3600x + 48G RAM + 1T NVMe

**Fuzz:** installed by `cargo install cargo-fuzz` hours before
**Command:** `make fuzz`
**Error:**
```bash
#36421  NEW    cov: 4429 ft: 8105 corp: 350/7112b lim: 38 exec/s: 22 rss: 722Mb L: 36/38 MS: 2 ChangeBinInt-PersAutoDict- DE: "\377\377\377\377"-
        NEW_FUNC[1/1]: 0x564a283e6fa0  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x59efa0) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
#36938  NEW    cov: 4435 ft: 8130 corp: 351/7154b lim: 43 exec/s: 22 rss: 726Mb L: 42/42 MS: 4 CopyPart-ChangeBit-CopyPart-CopyPart-
#36948  NEW    cov: 4435 ft: 8138 corp: 352/7195b lim: 43 exec/s: 22 rss: 726Mb L: 41/42 MS: 3 ChangeByte-ChangeByte-CopyPart-
#36954  NEW    cov: 4435 ft: 8144 corp: 353/7238b lim: 43 exec/s: 22 rss: 726Mb L: 43/43 MS: 4 EraseBytes-InsertRepeatedBytes-ShuffleBytes-ChangeBinInt-
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `"Err(NotIndexed)"`,
 right: `"Ok([Vertices([])])"`', fuzz_targets/compare.rs:549:17
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8996ea93b6e554148c4286e62b613f12a3ee505c/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/8996ea93b6e554148c4286e62b613f12a3ee505c/library/core/src/panicking.rs:64:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: compare::_::run
   5: rust_fuzzer_test_input
   6: std::panicking::try::do_call
   7: __rust_try
   8: LLVMFuzzerTestOneInput
   9: _ZN6fuzzer6Fuzzer15ExecuteCallbackEPKhm
  10: _ZN6fuzzer6Fuzzer6RunOneEPKhmbPNS_9InputInfoEbPb
  11: _ZN6fuzzer6Fuzzer16MutateAndTestOneEv
  12: _ZN6fuzzer6Fuzzer4LoopERSt6vectorINS_9SizedFileESaIS2_EE
  13: _ZN6fuzzer12FuzzerDriverEPiPPPcPFiPKhmE
  14: main
  15: <unknown>
  16: __libc_start_main
  17: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
==1370616== ERROR: libFuzzer: deadly signal
    #0 0x564a2853e911  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x6f6911) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #1 0x564a2884ca49  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0xa04a49) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #2 0x564a28837e85  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9efe85) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #3 0x7f58b0051f4f  (/usr/lib/libc.so.6+0x38f4f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #4 0x7f58b00a08eb  (/usr/lib/libc.so.6+0x878eb) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #5 0x7f58b0051ea7  (/usr/lib/libc.so.6+0x38ea7) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #6 0x7f58b003b53c  (/usr/lib/libc.so.6+0x2253c) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #7 0x564a295ddc76  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x1795c76) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #8 0x564a2847ff86  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x637f86) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #9 0x564a288320e3  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9ea0e3) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #10 0x564a295d2809  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x178a809) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #11 0x564a295d2588  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x178a588) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #12 0x564a295cf94b  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x178794b) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #13 0x564a295d2291  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x178a291) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #14 0x564a28483902  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x63b902) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #15 0x564a28483c3e  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x63bc3e) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #16 0x564a283e217d  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x59a17d) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #17 0x564a2882916f  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9e116f) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #18 0x564a2881dc19  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9d5c19) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #19 0x564a2882d06f  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9e506f) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #20 0x564a28832307  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9ea307) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #21 0x564a28831655  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9e9655) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #22 0x564a288383b9  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9f03b9) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #23 0x564a2883f9c5  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9f79c5) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #24 0x564a28840882  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9f8882) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #25 0x564a28841667  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x9f9667) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #26 0x564a288612cb  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0xa192cb) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #27 0x564a28484052  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x63c052) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)
    #28 0x7f58b003c78f  (/usr/lib/libc.so.6+0x2378f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #29 0x7f58b003c849  (/usr/lib/libc.so.6+0x23849) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #30 0x564a284b12a4  (/home/lab0/indradb-latest/lib/fuzz/target/x86_64-unknown-linux-gnu/release/compare+0x6692a4) (BuildId: e80228816f1114dcadee50251030af005dbef9b2)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 1 CrossOver-; base unit: 4e545f90f529340fdae8baa57777f22dfa12035f
0x3b,0x6c,0x6c,0x0,0x0,0xff,0xf7,0xff,0xff,0x49,0x49,0xff,0x49,0x49,0x49,0x49,0x49,0x4f,0x4e,0x49,0x49,0xf7,0x49,0x49,0x39,
;ll\000\000\377\367\377\377II\377IIIIIONII\367II9
artifact_prefix='/home/lab0/indradb-latest/lib/fuzz/artifacts/compare/'; Test unit written to /home/lab0/indradb-latest/lib/fuzz/artifacts/compare/crash-1684cda953e152514b888e5ee4f02d8664e93974
Base64: O2xsAAD/9///SUn/SUlJSUlPTklJ90lJOQ==

────────────────────────────────────────────────────────────────────────────────

Failing input:

        fuzz/artifacts/compare/crash-1684cda953e152514b888e5ee4f02d8664e93974

Output of `std::fmt::Debug`:

        [
            BulkInsert(
                [],
            ),
            IndexProperty(
                Identifier(
                    Identifier(
                        0x00006030000107e0 : "II",
                    ),
                ),
            ),
            Get(
                VertexWithPropertyValue(
                    VertexWithPropertyValueQuery {
                        name: Identifier(
                            Identifier(
                                0x00006030000107e0 : "II",
                            ),
                        ),
                        value: Null,
                    },
                ),
            ),
        ]

Reproduce with:

        cargo fuzz run compare fuzz/artifacts/compare/crash-1684cda953e152514b888e5ee4f02d8664e93974

Minimize test case with:

        cargo fuzz tmin compare fuzz/artifacts/compare/crash-1684cda953e152514b888e5ee4f02d8664e93974

────────────────────────────────────────────────────────────────────────────────

Error: Fuzz target exited with exit status: 77
make: *** [Makefile:19: fuzz] Error 1
[lab0@monitor indradb-latest]$ git log
commit 8f5bad2d62b0ddf9b90c9e38ebb211f20b8a3025 (HEAD)
Author: Yusuf Simonson <simonson@gmail.com>
Date:   Wed Feb 1 07:45:52 2023 -0500

    Added v4 warning

commit bc13dd486021dc2fe45dca889ba0fae83cb2cee4
```

--- a/lib/src/memory/datastore.rs
+++ b/lib/src/memory/datastore.rs
@@ -39,16 +39,6 @@ pub struct MemoryTransaction<'a> {
     path: Option<PathBuf>,
 }
 
-impl<'a> MemoryTransaction<'a> {
-    fn get_property_values(&self, name: Identifier) -> Result<&HashMap<Json, HashSet<IndexedPropertyMember>>> {
-        if let Some(container) = self.internal.property_values.get(&name) {
-            Ok(container)
-        } else {
-            Err(Error::NotIndexed)
-        }
-    }
-}
-
 impl<'a> Transaction<'a> for MemoryTransaction<'a> {
     fn vertex_count(&self) -> u64 {
         self.internal.vertices.len() as u64
@@ -99,13 +89,17 @@ impl<'a> Transaction<'a> for MemoryTransaction<'a> {
     }
 
     fn vertex_ids_with_property_value(&'a self, name: Identifier, value: &Json) -> Result<Option<DynIter<'a, Uuid>>> {
-        let container = self.get_property_values(name)?;
-        if let Some(sub_container) = container.get(value) {
-            let iter = Box::new(sub_container.iter().filter_map(move |member| match member {
-                IndexedPropertyMember::Vertex(id) => Some(Ok(*id)),
-                _ => None,
-            }));
-            Ok(Some(Box::new(iter)))
+        if let Some(container) = self.internal.property_values.get(&name) {
+            if let Some(sub_container) = container.get(value) {
+                let iter = Box::new(sub_container.iter().filter_map(move |member| match member {
+                    IndexedPropertyMember::Vertex(id) => Some(Ok(*id)),
+                    _ => None,
+                }));
+                Ok(Some(Box::new(iter)))
+            } else {
+                let iter = Vec::default().into_iter();
+                Ok(Some(Box::new(iter)))
+            }
         } else {
             Ok(None)
         }
@@ -155,15 +149,18 @@ impl<'a> Transaction<'a> for MemoryTransaction<'a> {
     }
 
     fn edges_with_property_value(&'a self, name: Identifier, value: &Json) -> Result<Option<DynIter<'a, Edge>>> {
-        let container = self.get_property_values(name)?;
-        if let Some(sub_container) = container.get(value) {
-            let iter = Box::new(sub_container.iter().filter_map(move |member| match member {
-                IndexedPropertyMember::Edge(edge) if self.internal.edges.contains(edge) => Some(edge),
-                _ => None,
-            }));
-            Ok(Some(Box::new(iter.map(|e| Ok(e.clone())))))
+        if let Some(container) = self.internal.property_values.get(&name) {
+            if let Some(sub_container) = container.get(value) {
+                let iter = Box::new(sub_container.iter().filter_map(move |member| match member {
+                    IndexedPropertyMember::Edge(edge) if self.internal.edges.contains(edge) => Some(edge),
+                    _ => None,
+                }));
+                Ok(Some(Box::new(iter.map(|e| Ok(e.clone())))))
+            } else {
+                Ok(None)
+            }
         } else {
-            Ok(None)
+            Err(Error::NotIndexed)
         }
     }
 

--- a/lib/src/tests/indexing.rs
+++ b/lib/src/tests/indexing.rs
@@ -275,6 +275,10 @@ pub fn should_query_indexed_edge_property_empty<D: Datastore>(db: &Database<D>) 
 pub fn should_get_vertex_with_property_value_empty<D: Datastore>(db: &Database<D>) {
     let property_name = models::Identifier::new("II").unwrap();
     db.index_property(property_name).unwrap();
-    let results = util::get_vertices(db, models::VertexWithPropertyValueQuery::new(property_name, ijson!(null))).unwrap();
+    let results = util::get_vertices(
+        db,
+        models::VertexWithPropertyValueQuery::new(property_name, ijson!(null)),
+    )
+    .unwrap();
     assert_eq!(results.len(), 0);
 }

--- a/lib/src/tests/indexing.rs
+++ b/lib/src/tests/indexing.rs
@@ -269,3 +269,12 @@ pub fn should_query_indexed_edge_property_empty<D: Datastore>(db: &Database<D>) 
     let result = util::get_edges(db, models::EdgeWithPropertyPresenceQuery::new(property_name)).unwrap();
     assert_eq!(result.len(), 0);
 }
+
+/// Tests for a regression found by the fuzzer:
+/// https://github.com/indradb/indradb/issues/278
+pub fn should_get_vertex_with_property_value_empty<D: Datastore>(db: &Database<D>) {
+    let property_name = models::Identifier::new("II").unwrap();
+    db.index_property(property_name).unwrap();
+    let results = util::get_vertices(db, models::VertexWithPropertyValueQuery::new(property_name, ijson!(null))).unwrap();
+    assert_eq!(results.len(), 0);
+}

--- a/lib/src/tests/macros.rs
+++ b/lib/src/tests/macros.rs
@@ -84,6 +84,7 @@ macro_rules! full_test_impl {
         define_test!(should_update_indexed_edge_property, $code);
         define_test!(should_query_indexed_vertex_property_empty, $code);
         define_test!(should_query_indexed_edge_property_empty, $code);
+        define_test!(should_get_vertex_with_property_value_empty, $code);
 
         // Properties
         define_test!(should_handle_vertex_properties, $code);

